python/event_generator.cpp

```diff
@@ -0,0 +1,262 @@
+#include <sstream>
+#include <string>
+
+#include <arbor/common_types.hpp>
+#include <arbor/event_generator.hpp>
+#include <arbor/schedule.hpp>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/pytypes.h>
+#include <pybind11/stl.h>
+
+namespace pyarb {
+
+// A Python shim that holds the information that describes an
+// arb::regular_schedule. This is wrapped in pybind11, and users constructing
+// a regular_schedule in python are manipulating this type. This is converted to
+// an arb::regular_schedule when a C++ recipe is created from a Python recipe.
+
+struct regular_schedule_shim {
+    using time_type = arb::time_type;
+
+    pybind11::object tstart = pybind11::none();
+    time_type dt = 0.;
+    pybind11::object tstop = pybind11::none();
+
+    regular_schedule_shim() = default;
+
+    regular_schedule_shim(time_type deltat):
+        tstart(pybind11::cast<time_type>(0.)),
+        dt(deltat)
+    {}
+
+    regular_schedule_shim(pybind11::object t0, time_type deltat, pybind11::object t1):
+        tstart(t0),
+        dt(deltat),
+        tstop(t1)
+    {}
+
+    arb::schedule schedule() const {
+        return arb::regular_schedule(tstart.is_none()? arb::terminal_time: pybind11::cast<time_type>(tstart), dt, tstop.is_none()? arb::terminal_time: pybind11::cast<time_type>(tstop));
```

Comment:
These will give hard errors if `tstart` and `tstop` can't be coverted to `time_type`.

e.g.
```
s = arb.regulard_schedule()
s.tstart = 'cat'
```
When the string is cast to float, and exception will be thrown, which will give the user a confusing error message. I have just noticed that the same thing will happen in the context. 

We might wrap the `pybind11::cast` call in a try-catch block, that throws an exception with a meaningful error message, such as "tstart must be either None or a non-negative value".

Reply:
When initializing `tstart` and `tstop` with `pybind::none()` as `pybind11:object` the constructors of the `regular_schedule` won't throw exceptions, since every `object` is legitimate. Only when constructing the `event_generator` it throws an exception where it is too late. So should we consider `arb::time_type` and setting `None` as default argument (allowed in default setting of args) which is then set to `arb::terminal_time`? That way python should throw a "wrong argument type" error instead.

Reply:
If we keep `pybind1::object`'s the try-catch block only hits in `schedule()` which is called in the `make_event_generator()`

```
    arb::schedule schedule() const {
        time_type t0;
        time_type t1;
        try {
            t0 = tstart.is_none()? arb::terminal_time: pybind11::cast<time_type>(tstart);
        }
        catch (...){
            std::cerr << "casting error: argument tstart must be either None or a non-negative value." << "\n";
        }
        try {
            t1 = tstop.is_none()? arb::terminal_time: pybind11::cast<time_type>(tstop);
        }
        catch (...){
            std::cerr << "casting error: argument tstop must be either None or a non-negative value." << "\n";
        }
        return arb::regular_schedule(t0, dt, t1);
    }
```

Reply:
Is there a way to include a try-catch in an initialization list?

Reply:
There is, it turns out!
https://en.cppreference.com/w/cpp/language/function-try-block

Reply:
There is no way to put try-catch in an initialization list.
Instead of `std::cerr`, I would throw a `std::runtime_error` with our clear message. The error message could be reduced to "regular schedule tstart must be either None or non-negative number."

You are right that it is not good to catch the problem "late" when converting from Python shim to a C++ `schedule`, because the error thrown at a different location to where the user set the invalid value.

How about writing a "setter" for the shim, that will check the type of the value being set by the Python user? The pybind11 docs show how to control access to properties via setters and getters. You could implement the setters and getters as lambdas.

https://pybind11.readthedocs.io/en/master/classes.html#instance-and-static-fields

Here is some pseudo code that just sets without checking: you could add the checking to the setter lambda:
```
    .def_readwrite("tstart",
        [](regular_schedule_shim& s, pybind11::object v) {s.tstart=v;},
        [](regular_schedule_shim& s) {return s.tstart;},                                                                                    
        "The delivery time of the first event in the sequence (in ms).")
```

Reply:
I misunderstood — I thought you were referring to member initialization.

Reply:
With 

```
.def_property("tstart", &regular_schedule_shim::get_tstart, &regular_schedule_shim::set_tstart,
            "The delivery time of the first event in the sequence (in ms).")
```

using 

```
void set_tstart(const pybind11::object v) {
        try {
            if (!v.is_none()) {
                tstart = v;
                if (tstart.cast<time_type>() < 0. or !pybind11::isinstance<time_type>(tstart) or !pybind11::isinstance<std::string>(pybind11::str(tstart)) )
                    throw std::runtime_error("regular schedule tstart must be either None or a non-negative value.");
            }
            else {
                tstart = v;
                if (!v.is_none())
                    throw std::runtime_error("regular schedule tstart must be either None or a non-negative value.");
            }
        }
        catch (std::runtime_error& e) {
            std::cout << e.what() << "\n";
        }
    };

const pybind11::object &get_tstart() const { return tstart; }
```

I still get 

```
>>> import arbor
>>> rs = arbor.regular_schedule()
>>> rs
<regular_schedule: tstart None, dt 0 ms, tstop None>
>>> rs.tstart=-1
regular schedule tstart must be either None or a non-negative value.
>>> rs.tstart=True
regular schedule tstart must be either None or a non-negative value.
>>> rs.tstart="cat"
Unable to cast Python instance to C++ type (compile in debug mode for details)
>>> rs.tstart='dog'
Unable to cast Python instance to C++ type (compile in debug mode for details)
```

Reply:
I think it is because the `cast<time_type>` throws an exception with the message `"Unable to cast Python instance to C++ type (compile in debug mode for details)"`, which is caught and then just forwarded.